app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt

```diff
@@ -0,0 +1,99 @@
+import com.google.firebase.firestore.GeoPoint
+import kotlinx.coroutines.tasks.await
+import java.io.Serializable
+import java.util.*
+
+data class Line(
+    val categoryRef: DocumentReference? = null,
+    val idCity: String = "",
+    val name: String = "",
+    val routePoints: List<GeoPoint> = listOf(),
+    val start: GeoPoint? = null,
+    val stops: List<GeoPoint> = listOf()
+) : Serializable {
+
+    companion object {
+        fun geoPointToLocation(data: GeoPoint): Location {
+            val newLocation = Location(LocationManager.NETWORK_PROVIDER)
+            newLocation.latitude = data.latitude
+            newLocation.longitude = data.longitude
+            return newLocation
+        }
+
+        fun geoPointListToLocationList(dataList: List<GeoPoint>): List<Location> {
+            return dataList.map { geoPointToLocation(it) }
+        }
+    }
+
+    suspend fun lineToLinePath(): LinePath {
+        val routePoints = geoPointListToLocationList(routePoints)
+        val start = start?.let { geoPointToLocation(it) }
+        val stops = geoPointListToLocationList(stops)
+        var category: DocumentSnapshot?
+        var categoryName = ""
+        categoryRef?.let { docRef ->
+            category = docRef.get().await()
+            category?.let {
+                val currLang = Locale.getDefault().isO3Language
+                categoryName =
+                    if (currLang == "spa") it.toObject(LineCategories::class.java)?.nameEsp ?: ""
+                    else it.toObject(LineCategories::class.java)?.nameEng ?: ""
+            }
+        }
+        return LinePath(name, categoryName, routePoints, start, stops)
+    }
+}
+
+data class LinePath(
+    val name: String = "",
+    val category: String = "",
+    val routePoints: List<Location> = listOf(),
+    val start: Location? = null,
+    val stops: List<Location> = listOf()
+) {
+    companion object {
+        fun getOneRouteLine(line: LinePath, nearestStopToDestination: Location, nearestStopToOrigin: Location): AvailableTransport? {
+            val indexOrigin = line.stops.indexOf(nearestStopToOrigin)
+            val indexDestination = line.stops.indexOf(nearestStopToDestination)
+
+            if (indexOrigin < indexDestination) {
+                val newStops = line.stops.slice(indexOrigin..indexDestination)
+
+                val indexOriginPoint = getIndexOfFromLocationList(nearestStopToOrigin, line.routePoints)
+                val indexDestinationPoint = getIndexOfFromLocationList(nearestStopToDestination, line.routePoints)
+
+                val newRoutePoints = line.routePoints.slice(indexOriginPoint..indexDestinationPoint)
+                val newLine = LinePath(line.name, line.category, newRoutePoints, line.start, newStops)
+                return AvailableTransport(null, mutableListOf(newLine))
+            }
+            return null
+        }
+
+        fun getSubLine(line: LinePath, nearestStop: Location, isForDestination: Boolean): LinePath {
```

Comment:
It's not so clear the name of the function with the method, can you add a comment?

Reply:
added comments for clarification